### PR TITLE
Improve and extend GIAS persistence

### PR DESCRIPTION
**Pull request by peterdavidhamilton**
## Changes in this PR

- move persisting responsibility to a dedicated service
- control lazy iteration with a configurable variable
- this should make logging and reporting success/failure easier

## TODO

https://trello.com/c/4tDDmqA3

### CODE REVIEW

File: lib/school/record_keeper.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  # @note Changes to school records may impact active cases.
+  #
+  class RecordKeeper
+    # @param records [Array<Hash>]
+    def call(records)
+      # OPTIMIZE: opportunity to wrap in a transaction block
+      records.map do |record|
+        if legacy_record?(record)
+          # puts "School URN: #{record[:urn]} is already closed"
+          next
```
Comment (kennyevil) [inline lib/school/record_keeper.rb:14]: This could be a one liner if nothing is going in the block ie ```next if legacy_record?(record)```

**Review by kennyevil — APPROVED** [review]
This looks great. Only one incredibly minor quibble but that's it.

File: db/schema.rb [hunk 2]
```diff
@@ -45,7 +45,7 @@
     t.datetime "created_at", precision: 6, default: -> { "CURRENT_TIMESTAMP" }, null: false
     t.datetime "updated_at", precision: 6, default: -> { "CURRENT_TIMESTAMP" }, null: false
     t.integer "journeys_count"
-    t.string "slug"
+    t.string "slug", null: false
```
Comment (kennyevil) [inline db/schema.rb:48]: Remove this

File: lib/school/record_keeper.rb [hunk 3]
```diff
@@ -52,5 +59,11 @@ def legacy_record?(record)
     def type(record)
       Support::EstablishmentType.find_by(code: record[:establishment_type][:code])
     end
+
```
Comment (kennyevil) [inline lib/school/record_keeper.rb:62]: Add comment on why this method is built this way

File: lib/school/schema.rb [hunk 4]
```diff
@@ -13,6 +13,7 @@ class Schema < AppSchema
       required(:urn).filled(:integer)
       required(:uprn).filled(:integer)
       required(:crn).maybe(:string)
+      required(:rsc_region).maybe(:string)
```
Comment (kennyevil) [inline lib/school/schema.rb:16]: Change to filled

File: db/migrate/20211101171222_add_extra_gias_fields_to_organisations.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class AddExtraGiasFieldsToOrganisations < ActiveRecord::Migration[6.1]
+  def change
+    change_table :support_organisations, bulk: true do |t|
+      t.column :ukprn, :string
+      t.column :telephone_number, :string
+      t.column :local_authority, :jsonb
+      t.column :open_date, :datetime
```
Comment (kennyevil) [inline db/migrate/20211101171222_add_extra_gias_fields_to_organisations.rb:7]: Change to opened_date

File: lib/school/record_keeper.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    def type(record)
+      Support::EstablishmentType.find_by(code: record[:establishment_type][:code])
+    end
+
+    # Some organisations do not have this information on their record and .parse does
+    # not like dealing with nil or strings that it can't parse. This method helps
+    # to deal with that. The regex is based on the format of the dates used in the file
+    # @param opened_date [String]
+    #
+    # @return [Time] or nil
```
Comment (peterdavidhamilton) [inline lib/school/record_keeper.rb:68]: if you run `yard` see if it comes out as expected - `[Time, nil]`

File: lib/school/record_keeper.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    #
+    # @return [Support::EstablishmentType]
+    def type(record)
+      Support::EstablishmentType.find_by(code: record[:establishment_type][:code])
+    end
+
+    # Some organisations do not have this information on their record and .parse does
+    # not like dealing with nil or strings that it can't parse. This method helps
+    # to deal with that. The regex is based on the format of the dates used in the file
+    # @param opened_date [String]
```
Comment (peterdavidhamilton) [inline lib/school/record_keeper.rb:66]: Can you append an example of the format used in the file

Comment (peterdavidhamilton) [reply lib/school/record_keeper.rb:66]: `# @param opened_date [String] xxxxxx`

File: spec/lib/self-serve/school/record_keeper_spec.rb [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        before { create(:support_organisation, urn: new_record[:urn]) }
+
+        it "returns false" do
+          expect(record_keeper.send(:legacy_record?, new_record)).to eq(false)
+        end
+      end
+    end
+  end
+
+  describe "#parse_opened_date" do
```
Comment (peterdavidhamilton) [inline spec/lib/self-serve/school/record_keeper_spec.rb:245]: You're testing code internals, implementation details, but we could assert this using a fixture with nil for example.

**Comment by kennyevil** [comment]
[Add extra fields to GIAS records](https://trello.com/c/4tDDmqA3/259-add-extra-fields-to-gias-records)

File: spec/lib/self-serve/school/record_keeper_spec.rb [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        record_keeper.call([legacy_record, new_record])
+
+        saved_keys.each do |key|
+          expect(Support::Organisation.last.send(key).blank?).to be(false)
+        end
+      end
+    end
+  end
+
+  describe "legacy_record" do
```
Comment (peterdavidhamilton) [inline spec/lib/self-serve/school/record_keeper_spec.rb:211]: As we expect the records to evolve as we define the tables, fields and relationships, I'd say it is a bit too early to test this but I'd argue that as it is private, we shouldn't anyway.

**Comment by peterdavidhamilton** [comment]
[Improve GIAS efficiency](https://trello.com/c/50r3MGyg/248-improve-gias-efficiency)

File: app/services/support/seed_schools.rb [hunk 10]
```diff
@@ -1,70 +1,48 @@
...
-    # @return [Array<Hash>] loaded from CSV file or updated live
-    def dataset
-      data ? local_dataset.call : live_dataset.call
-    end
+    # Filter records that do match these criteria
+    #
+    # @!attribute [r] filter
+    #   @return [Hash] "column header" => [integer, values]
+    option :filter, ::Types::Strict::Hash, reader: :private, default: proc {
+      { "TypeOfEstablishment (code)" => EstablishmentType.all.map(&:code) }
```
Comment (ryantk) [inline app/services/support/seed_schools.rb:30]: Pluck is more performant than all.map.x
https://apidock.com/rails/ActiveRecord/Calculations/pluck

File: lib/school/information.rb [hunk 11]
```diff
@@ -1,63 +1,92 @@
...
       CSV.foreach(data_file.path, headers: true, encoding: "ISO-8859-1:UTF-8").lazy
     end
 
-    # @return [Array<Hash>] filter criteria matches the row
-    def filtered_data
-      tuples.select do |row|
+    # Converts "01" to 1
+    #
+    # @return [Enumerator::Lazy] filter rows by codes
+    def filtered_rows
```
Comment (ryantk) [inline lib/school/information.rb:87]: One thing I would query about this code is School::Information (according to comments) knows a lot about the fact it is being filtered by code, yet this is being passed in as a generic parameter so the assumption is that this filter can change dynamically. 

I would either remove reference to code in comments, or hardcode "filter".

I will not request changes for this as it is quite subtle and up for debate.

File: lib/school/record_keeper.rb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+          # The "Organisation" model is a starting point, but in addition an
+          # "Establishment" model has been proposed.
+          #
+          Support::Organisation.find_or_initialize_by(urn: record[:urn]).tap do |org|
+            org.establishment_type_id = type(record).id                         # uuid
+            org.name = record[:school][:name]                                   # string
+            org.address = record[:school][:address]                             # jsonb
+            org.contact = record[:school][:head_teacher]                        # jsonb
+            org.phase = record[:school][:phase][:code]                          # integer
+            org.gender = record[:school][:gender][:code]                        # integer
```
Comment (ryantk) [inline lib/school/record_keeper.rb:29]: Dig is quite nice in this scenario.
https://apidock.com/rails/ActiveSupport/HashWithIndifferentAccess/dig

**Review by ryantk — APPROVED** [review]
This looks good as a whole. I have made some minor comments but they are not big enough for me to request changes, just something to consider.

**Review by d-a-v-e — APPROVED** [review]

File: lib/school/record_keeper.rb (hunk 12 above)
Comment (peterdavidhamilton) [reply lib/school/record_keeper.rb:29]: Good shout @ryantk, at the moment the structure of the transformer output is kind of generic but as we solidify what attributes go where I'd like to see the transformer used to create that structure so we just splat hashes into `create`... or whatever.

**Pull request merged**